lib/pos_network.go

```diff
@@ -404,3 +404,119 @@ func DecodeAggregatedBLSSignature(rr io.Reader) (*AggregatedBLSSignature, error)
 
 	return &sig, nil
 }
+
+// TimeoutAggregateQuorumCertificate is an aggregation of timeout messages from 2/3rds
+// of all validators, weighted by stake, which indicates that these validators want to
+// time out a particular view.
+//
+// When validators want to time a view, they send their high QCs to the block proposer
+// who builds an aggregate QC extending the chain from the highest QC that it received
+// from all validators who timed out. To prove that it has selected the highest QC, the
+// proposer also includes a list of the high QC views that each validator has sent.
+type TimeoutAggregateQuorumCertificate struct {
+
+	// The view that the block proposers has produced a timeout QC for.
+	TimedOutView uint64
+
+	// This is the highest QC that the block proposer received from any validator who
+	// has timed out for the current view.
+	ValidatorsHighQC *QuorumCertificate
+
+	// Here we include a list of the HighQC.View values we got from each of the
+	// validators in the ValidatorsTimeoutHighQCViews field. In addition, for each
+	// unique HighQC.View value we received, we combine all the partial signatures
+	// for that HighQC.View into a single BLSMultiSignature.
+	//
+	// Note: The aggregated signature is made up of partial signatures for all
+	// present validators, each of whom signed a payload with the pair
+	// (current view, the validator's local HighQC.View).
+	ValidatorsTimeoutHighQCViews         []uint64
+	ValidatorsTimeoutAggregatedSignature *AggregatedBLSSignature
+}
+
+// Performs a deep equality check between two TimeoutAggregateQuorumCertificate, and
+// returns true if the values of the two are identical.
+func (aggQC *TimeoutAggregateQuorumCertificate) Eq(
+	other *TimeoutAggregateQuorumCertificate,
+) bool {
+	if aggQC == nil && other == nil {
+		return true
+	}
+
+	if (aggQC == nil) != (other == nil) {
+		return false
+	}
+
+	if len(aggQC.ValidatorsTimeoutHighQCViews) != len(other.ValidatorsTimeoutHighQCViews) {
+		return false
+	}
+
+	for i := 0; i < len(aggQC.ValidatorsTimeoutHighQCViews); i++ {
+		if aggQC.ValidatorsTimeoutHighQCViews[i] != other.ValidatorsTimeoutHighQCViews[i] {
+			return false
+		}
+	}
+
+	return aggQC.TimedOutView == other.TimedOutView &&
+		aggQC.ValidatorsHighQC.Eq(other.ValidatorsHighQC) &&
+		aggQC.ValidatorsTimeoutAggregatedSignature.Eq(other.ValidatorsTimeoutAggregatedSignature)
```

Comment:
Is it more efficient to perform this check before looping over the aggQC.ValidatorsTimeoutHighQCViews?

Reply:
Hmm, the code reads a bit more cleanly with your suggestion. Not much difference in terms of efficiency though since `aggQC.ValidatorsTimeoutAggregatedSignature.Eq` is an O(n) check. I've updated the code to reorder the checks